spec/models/test_track/session_spec.rb

```diff
@@ -422,6 +422,60 @@ def current_clown; end
     end
   end
 
+  describe "#visitor_dsl_for" do
+    before do
+      allow(TestTrack::Remote::Visitor).to receive(:from_identifier).and_call_original
+    end
+
+    context "when the controller's authenticated resource matches the identity" do
+      before do
+        allow(controller).to receive(:current_clown).and_return(identity)
+      end
+
+      it "doesn't fetch a remote visitor" do
+        expect(subject.visitor_dsl_for(identity)).to be_a TestTrack::VisitorDSL
+
+        expect(TestTrack::Remote::Visitor).not_to have_received(:from_identifier)
+      end
+    end
+
+    context "when the controller's authenticated resource does not match the identity" do
+      let(:other_identity) { Clown.new(id: 9876) }
+
+      before do
+        allow(controller).to receive(:current_clown).and_return(other_identity)
+      end
+
+      it "fetches a remote visitor" do
+        expect(subject.visitor_dsl_for(identity)).to be_a TestTrack::VisitorDSL
+
+        expect(TestTrack::Remote::Visitor).to have_received(:from_identifier).with('clown_id', 1234)
+      end
+    end
+  end
+
+  describe "#visitors_by_identity" do
+    let(:identity) { double(test_track_identifier_type: "foo_user_id", test_track_identifier_value: "123") }
+
+    before do
+      allow(TestTrack::Remote::Visitor).to receive(:from_identifier).and_call_original
+    end
+
+    it "fetches a remote visitor on demand given an identity" do
+      expect(subject.visitors_by_identity[identity]).to be_a TestTrack::Visitor
+
+      expect(TestTrack::Remote::Visitor).to have_received(:from_identifier).with("foo_user_id", "123")
+    end
+
+    it "only fetches a remote visitor once for the same identity" do
+      subject.visitors_by_identity[identity]
+      subject.visitors_by_identity[identity]
+
+      expect(TestTrack::Remote::Visitor).to have_received(:from_identifier).with("foo_user_id", "123").exactly(:once)
```

Comment:
It's kinda nice to test this separately, but I don't think it'd be very gnarly to test this through `visitor_dsl_for`. Of course the real pro move would be to extract to another class. But I'd like to do that as I get rid of the IdentityCollection class which is really confusing.